### DISCUSSION TITLE
How do I integrate SYCL code into an existing GCC/C++ application with minimum disturbance?

### DISCUSSION THREAD

**Discussion by biergaizi** [AdaptiveCpp/AdaptiveCpp#1085] [Q&A] [answered]
I'm working on implementing an SYCL-accelerated kernel in a preexisting C++/GCC-based codebase. Early experiments showed that SYCL is promising. However, I don't how to integrate SYCL code into the existing codebase in a way with minimum disturbance.

The most straightforward solutions are:

1. Rebuilding the entire C++ codebase with OpenSYCL toolchain. However, it means that the entire application is now recompiled by clang instead of GCC, this affects 100% of the object code and is undesirable. For example, the performance characteristic of auto-vectorized CPU code may change.
2. Implementing the SYCL portion of the code as a shared library, building it using OpenSYCL or other SYCL toolchains, and loading the library from the main application dynamically. This is much better but can be troublesome, since the application is currently not designed for loading different kernels as shared libraries. It uses conventional C++ classes and subclasses for different kernels.

Since the new SYCL code is implemented as C++ subclasses, thus, in the ideal world, the best work flow would be:

1. Build the non-SYCL C++ classes to object code with the existing GCC toolchain.
2. Build the SYCL C++ classes to object code with OpenSYCL/clang toolchain.
3. Link object code of both GCC/C++ and OpenSYCL output to an executable. 

Is it possible to do so using OpenSYCL? I'm afraid that it's not supported - from my experience of sysadmining, linking outputs from different compilers is generally not a good idea due to potential conflicts, and I suspect SSCP won't work even if it's technically possible.

If it's not supported, what is your recommendation on integrating SYCL code into an existing application?

Thanks.

**Comment by illuhad** [comment, marked as answer]
We are not supporting this explicitly, but this is also because it's not really the responsibility of Open SYCL: For your desired use case, the usual ABI compatibility guarantees between clang and gcc apply, and I don't think we do anything that alters this.

Normally clang and gcc work very well together, as long as they use the same `libstdc++` (clang by default just uses the one on the system) and neither of them is ancient (like say gcc 4.8.5 on some Red Hat systems).

EDIT: My recommendation is to also let Open SYCL handle the linking process, so that it can automatically link required runtime and backend libraries. So you'd need to compile the SYCL classes using Open SYCL, and then the final linking step of the application.
